### PR TITLE
#157964345 Build out CRUD API endpoints

### PR DESCRIPTION
**Pull request by denniswanjiru**
#### What does this PR do?
Building out the CRUD API endpoints

#### Description of Task to be completed?
METHOD | ENDPOINT | DESCRIPTION 
------------ | ------------- | -------------
GET | /api/v1/users/requests/    | Fetch all the requests resources
GET | /api/v1/users/request/<int:request_id>/   | Fetch a single request resource
POST | /api/v1/users/requests/    | Create a new request
PUT | /api/v1/users/request/<int:request_id>/    | Update a specific request
DELETE | /api/v1/users/request/<int:request_id>/    | Delete a specific request
POST | /api/v1/users/auth/signup/    | Handle user registration
POST | /api/v1/users/auth/signin/    | Handle user login

#### How should this be manually tested?
1. Clone the branch into your local machine as follows
2. Run the following commands on your terminal 

```
git clone -b ch-setup-api-env-157930602 git@github.com:DennisWanjiru/maintenance-tracker.git 

cd maintenance-tracker/

python3 -m virtualenv env -p python3.6

source env/bin/activate

pip install -r requirements.txt

python run.py
```
Test the above methods in postman

#### Any background context you want to provide?
Written in flask-restful

#### What are the relevant pivotal tracker stories?
[#157964345](https://www.pivotaltracker.com/story/show/157964345)

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -1,3 +1,113 @@
...
+parser.add_argument('description')
+
+
+def find_request(request_id):
+    """ Find a specific request resource based off the id """
+    return [_request for _request in requests_store if _request['request_id'] == request_id]
+
+
+class RequestList(Resource):
+    def get(self):
```
Comment (victoriaaoka) [inline app/__init__.py]: How do you ensure that users can only view their own requests, and the admin can view all the requests?

**Review by victoriaaoka — COMMENTED** [review]
Good implementation so far, but you need to handle edge cases.

File: app/__init__.py (hunk 1 above)
Comment (denniswanjiru) [reply app/__init__.py]: Okay working on the two changes.

File: app/__init__.py [hunk 2]
```diff
@@ -1,3 +1,114 @@
...
+
+
+class UserRegistration(Resource):
+    def post(self):
+        data = request.get_json()
+        username_taken = [
+            username for username in users if data["username"] in users]
+        emails = [users[user]["email"] for user in users]
+
+        if username_taken:
```
Comment (victoriaaoka) [inline app/__init__.py:88]: Do you have tests for this?

File: tests/test_requests_resource.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                "password": "mysupersecret",
+                "confirm_password": "mysupersecret",
+            },
+            "creds": {
+                "username": "user1",
+                "password": "mysupersecret"
+            }
+        }
+
+    # def register_user(self):
```
Comment (victoriaaoka) [inline tests/test_requests_resource.py]: Do you think this comments are at the right place?

File: tests/test_requests_resource.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        """ Test a resource can be successfully updated """
+        response = self.client.post(
+            '/api/v1/users/requests/',
+            data=json.dumps(self.data["dummy_request"]),
+            content_type=("application/json")
+        )
+        self.assertEqual(response.status_code, 201)
+
+        response = self.client.put(
+            '/api/v1/users/request/1/', data=json.dumps(dict(location="New title")), content_type=("application/json")
```
Comment (victoriaaoka) [inline tests/test_requests_resource.py]: Make this line shorter like the rest. This increases readability and remember PEP 8 styling :)

File: tests/test_auth.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import unittest
+import json
+
+
+class TestRequestResource(unittest.TestCase):
+    """ A class to perform tests for the Request Resource """
+
+    def setUp(self):
+        """
+        Setup the app to be in testing mode and make it able to use test client
```
Comment (victoriaaoka) [inline tests/test_auth.py]: One-liner docstrings have the opening/closing """ on the same line.

**Review by victoriaaoka — COMMENTED** [review]
- Add docstrings to all the functions. Some have them and some don't.
- Always use PEP 8 styling.

File: app/__init__.py (hunk 2 above)
Comment (denniswanjiru) [reply app/__init__.py:88]: Yess I have tests for this but the one for registering users is misbehaving. I'm trying to fix it

**Pull request merged**